README.md

# Twitmo <img src="man/figures/hexSticker.png" width="160px" align="right"/>

<!-- badges: start -->

[![](https://www.r-pkg.org/badges/version/Twitmo?color=green)](https://cran.r-project.org/package=Twitmo)
[![R-CMD-check](https://github.com/abuchmueller/Twitmo/workflows/R-CMD-check/badge.svg)](https://github.com/abuchmueller/Twitmo/actions)
[![License:
MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

<!-- badges: end -->

The goal of `Twitmo` is to facilitate topic modeling in R with Twitter
data. `Twitmo` provides a broad range of methods to sample, pre-process
and visualize contents of geo-tagged tweets to make modeling the public
discourse easy and accessible.

## Common questions

### Can I use `Twitmo` for pseudo-document pooling if I already sampled data earlier from Twitter without `Twitmo`?

Yes, this is possible in the Github version of `Twitmo.` You can use
`pool_tweets()` on any data frame, that has a ‘text’ and a ‘hashtags’
columns that are also named that way. Any additional columns you might
have, can additionally be used as document meta-data in a STM (see
below).

### Can I use `Twitmo` to model topical prevalence over time?

`Twitmo` has no built-in methods for this purpose, however slicing your
data time wise and fitting multiple LDA models then comparing topical
prevalence over time can be accomplished with `Twitmo` in conjunction
with `ggplot2`.

## Installation

## Important Note for **NEW** users

If you are using `Twitmo` for the first time, you might not already have
`rtweet` installed. If you have `rtweet` version \>= 1.0.0 installed,
you will not be able to use certain parts of `Twitmo`, like
parsing/loading tweets because of breaking changes in `rtweet`. Since
CRAN, by default, only distributes the latest version of a package and R
does not respect upper boundaries on dependencies I am currently working
on a solution. **You make sure you have the correct version of `rtweet`
installed by running**

``` r
## install remotes package if it's not already
if (!requireNamespace("devtools", quietly = TRUE)) {
  install.packages("devtools")
}

devtools::install_version("rtweet", version = "0.7.0", repos = "http://cran.us.r-project.org")
```

You can install `Twitmo` from CRAN with:

``` r
install.packages("Twitmo")
```

or install from Github where the correct version of `rtweet` will
automatically be installed.

You can install `Twitmo` from Github with:

``` r
## install remotes package if it's not already
if (!requireNamespace("remotes", quietly = TRUE)) {
  install.packages("remotes")
}

## install dev version of Twitmo from github
remotes::install_github("abuchmueller/Twitmo")
```

**Note**: Installing from Github may require you to have Rtools on your
system.

-   [Windows](https://cran.r-project.org/bin/windows/Rtools/ "Rtools for Windows (CRAN)")

-   [macOS](https://thecoatlessprofessor.com/programming/cpp/r-compiler-tools-for-rcpp-on-macos/ "Rtools for macOS")

## Collecting geo-tagged tweets

Make sure you have a regular Twitter Account before start to sample your
tweets.

``` r
# Live stream tweets from the UK for 30 seconds and save to "uk_tweets.json" in current working directory
get_tweets(method = 'stream', 
           location = "GBR", 
           timeout = 30, 
           file_name = "uk_tweets.json")

# Use your own bounding box to stream US mainland tweets
get_tweets(method = 'stream', 
           location = c(-125, 26, -65, 49), 
           timeout = 30,
           file_name = "tweets_from_us_mainland.json")
```

## Load your tweets from a json file into a data frame

A small sample with raw tweets is included in the package. Access via:

``` r
raw_path <- system.file("extdata", "tweets_20191027-141233.json", package = "Twitmo")
mytweets <- load_tweets(raw_path)
#>  Found 167 records... Found 193 records... Imported 193 records. Simplifying...
```

## Pool tweets into long pseudo-document

``` r
pool <- pool_tweets(mytweets)
#> 
#> 193 Tweets total
#> 158 Tweets without hashtag
#> Pooling 35 Tweets with hashtags #
#> 56 Unique hashtags total
#> Begin pooling ...Done
pool.corpus <- pool$corpus
pool.dfm <- pool$document_term_matrix
```

## Find optimal number of topics

``` r
find_lda(pool.dfm)
```

![](man/figures/README-ldatuner-1.png)<!-- -->

## Fitting a LDA model

``` r
model <- fit_lda(pool.dfm, n_topics = 7)
```

## View most relevant terms for each topic

``` r
lda_terms(model)
#>      Topic.1 Topic.2   Topic.3   Topic.4       Topic.5                Topic.6 Topic.7
#> 1        bio    last     today     paola tenrestaurant                  music  church
#> 2       link    meet    laurel      says         crazy                   time   today
#> 3      click  people      glen     puppy       covered               birthday morning
#> 4        job   first  trailing  downtown        waffle              girlhappy   place
#> 5        see     big      oaks knoxville        sooooo          birthdaytasha    life
#> 6        can  always   tuscany      like          life                morning     day
#> 7  recommend    love        ii        us     democrats                  early     see
#> 8     anyone   night    design    season          jeff             photoshoot  sunday
#> 9      great     fun perfectly    theres      sessions shamarathemodelscruggs    yall
#> 10      care    good     sized   nothing     beautiful               allwomen    word
```

or which hashtags are heavily associated with each topic

``` r
lda_hashtags(model)
#>                      Topic
#> mood                     7
#> motivate                 5
#> healthcare               1
#> mrrbnsnathome            5
#> newyork                  5
#> breakfast                5
#> thisismyplace            7
#> p4l                      7
#> chinup                   4
#> sundayfunday             4
#> saintsgameday            4
#> instapuppy               4
#> woof                     4
#> tailswagging             4
#> tickfire                 6
#> msiclassic               4
#> nyc                      2
#> about                    2
#> joethecrane              2
#> government               1
#> ladystrut19              6
#> ladystrutaccessories     6
#> smartnews                5
#> sundaythoughts           7
#> sf100                    6
#> openhouse                3
#> springtx                 3
#> labor                    1
#> norfolk                  1
#> oprylandhotel            2
#> pharmaceutical           3
#> easthanover              2
#> sales                    2
#> scryingartist            5
#> beautifulskyz            5
#> knoxvilletn              4
#> downtownknoxville        4
#> heartofservice           6
#> youthmagnet              6
#> youthmentor              6
#> bonjour                  4
#> trump2020                7
#> spiritchat               4
#> columbia                 1
#> newcastle                6
#> oncology                 1
#> nbatwitter               7
#> detroit                  1
```

## Inspecting LDA distributions

Check the distribution of your LDA Model with

``` r
lda_distribution(model)
#>                         V1    V2    V3    V4    V5    V6    V7
#> mood                 0.001 0.001 0.001 0.001 0.001 0.001 0.995
#> motivate             0.001 0.001 0.001 0.001 0.994 0.001 0.001
#> healthcare           0.996 0.001 0.001 0.001 0.001 0.001 0.001
#> mrrbnsnathome        0.002 0.002 0.002 0.002 0.989 0.002 0.002
#> newyork              0.002 0.002 0.002 0.002 0.989 0.002 0.002
#> breakfast            0.002 0.002 0.002 0.002 0.989 0.002 0.002
#> thisismyplace        0.001 0.001 0.001 0.001 0.001 0.001 0.994
#> p4l                  0.001 0.001 0.001 0.001 0.001 0.001 0.994
#> chinup               0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> sundayfunday         0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> saintsgameday        0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> instapuppy           0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> woof                 0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> tailswagging         0.004 0.004 0.004 0.978 0.004 0.004 0.004
#> tickfire             0.001 0.001 0.001 0.001 0.001 0.996 0.001
#> msiclassic           0.001 0.001 0.001 0.994 0.001 0.001 0.001
#> nyc                  0.001 0.996 0.001 0.001 0.001 0.001 0.001
#> about                0.001 0.996 0.001 0.001 0.001 0.001 0.001
#> joethecrane          0.001 0.996 0.001 0.001 0.001 0.001 0.001
#> government           0.995 0.001 0.001 0.001 0.001 0.001 0.001
#> ladystrut19          0.001 0.001 0.001 0.001 0.001 0.996 0.001
#> ladystrutaccessories 0.001 0.001 0.001 0.001 0.001 0.996 0.001
#> smartnews            0.001 0.001 0.001 0.001 0.997 0.001 0.001
#> sundaythoughts       0.000 0.000 0.000 0.000 0.000 0.000 0.997
#> sf100                0.001 0.001 0.001 0.001 0.001 0.995 0.001
#> openhouse            0.000 0.000 0.998 0.000 0.000 0.000 0.000
#> springtx             0.000 0.000 0.998 0.000 0.000 0.000 0.000
#> labor                0.995 0.001 0.001 0.001 0.001 0.001 0.001
#> norfolk              0.995 0.001 0.001 0.001 0.001 0.001 0.001
#> oprylandhotel        0.001 0.995 0.001 0.001 0.001 0.001 0.001
#> pharmaceutical       0.268 0.001 0.729 0.001 0.001 0.001 0.001
#> easthanover          0.001 0.996 0.001 0.001 0.001 0.001 0.001
#> sales                0.001 0.996 0.001 0.001 0.001 0.001 0.001
#> scryingartist        0.001 0.001 0.001 0.001 0.996 0.001 0.001
#> beautifulskyz        0.001 0.001 0.001 0.001 0.996 0.001 0.001
#> knoxvilletn          0.001 0.001 0.001 0.994 0.001 0.001 0.001
#> downtownknoxville    0.001 0.001 0.001 0.994 0.001 0.001 0.001
#> heartofservice       0.003 0.003 0.003 0.003 0.003 0.983 0.003
#> youthmagnet          0.003 0.003 0.003 0.003 0.003 0.983 0.003
#> youthmentor          0.003 0.003 0.003 0.003 0.003 0.983 0.003
#> bonjour              0.001 0.001 0.001 0.994 0.001 0.001 0.001
#> trump2020            0.001 0.001 0.001 0.001 0.001 0.001 0.994
#> spiritchat           0.001 0.001 0.001 0.996 0.001 0.001 0.001
#> columbia             0.996 0.001 0.001 0.001 0.001 0.001 0.001
#> newcastle            0.001 0.001 0.001 0.001 0.001 0.996 0.001
#> oncology             0.995 0.001 0.001 0.001 0.001 0.001 0.001
#> nbatwitter           0.000 0.000 0.000 0.000 0.000 0.000 0.997
#> detroit              0.995 0.001 0.001 0.001 0.001 0.001 0.001
```

# Filtering tweets

Sometimes you can build better topic models by blacklisting or
whitelisting certain keywords from your data. You can do this with a
keyword dictionary using the `filter_tweets()` function. In this example
we exclude all tweets with “football” or “mood” in them from our data.

``` r
mytweets %>% dim()
#> [1] 193  92
filter_tweets(mytweets, keywords = "football,mood", include = FALSE) %>% dim()
#> [1] 183  92
```

Analogously if you want to run your collected tweets through a whitelist
use

``` r
mytweets %>% dim()
#> [1] 193  92
filter_tweets(mytweets, keywords = "football,mood", include = TRUE) %>% dim()
#> [1] 10 92
```

# Fitting a structural topic model (STM)

Structural topic models can be fitted with additional external
covariates. In this example we metadata that comes with the tweets such
as retweet count. This works with parsed unpooled tweets. Pre-processing
and fitting is done with one function.

``` r
stm_model <- fit_stm(mytweets, n_topics = 7, xcov = ~ retweet_count + followers_count + reply_count + quote_count + favorite_count,
                     remove_punct = TRUE,
                     remove_url = TRUE,
                     remove_emojis = TRUE,
                     stem = TRUE,
                     stopwords = "en")
```

STMs can be inspected via

``` r
summary(stm_model)
#> A topic model with 7 topics, 137 documents and a 324 word dictionary.
#> Topic 1 Top Words:
#>       Highest Prob: last, today, time, love, show, week, just 
#>       FREX: love, show, week, give, whole, today, alway 
#>       Lift: alway, band, fan, give, love, miss, monster 
#>       Score: stop, love, last, today, week, give, show 
#> Topic 2 Top Words:
#>       Highest Prob: see, job, bio, link, click, can, season 
#>       FREX: bio, link, click, hire, latest, follow, downtown 
#>       Lift: bio, link, allen, area, develop, downtown, embarrass 
#>       Score: area, bio, link, job, click, hire, see 
#> Topic 3 Top Words:
#>       Highest Prob: like, one, year, want, think, read, win 
#>       FREX: year, way, around, parti, ppl, one, match 
#>       Lift: around, father, parti, ppl, pretti, start, throughout 
#>       Score: parti, one, year, like, way, ppl, match 
#> Topic 4 Top Words:
#>       Highest Prob: happen, life, right, now, say, need, hear 
#>       FREX: need, keep, action, support, work, right, hear 
#>       Lift: ’ve, action, difficult, footbal, head, keep, liter 
#>       Score: support, happen, say, now, need, life, action 
#> Topic 5 Top Words:
#>       Highest Prob: trump, know, peopl, game, best, presid, look 
#>       FREX: trump, know, care, got, intellig, isi, best 
#>       Lift: baghdadi, better, bin, fail, grow, person, sport 
#>       Score: isi, know, trump, care, intellig, presid, truth 
#> Topic 6 Top Words:
#>       Highest Prob: get, help, come, take, place, team, can 
#>       FREX: help, pleas, hors, colleg, communiti, made, question 
#>       Lift: hors, pleas, anytim, awesom, benefit, colleg, communiti 
#>       Score: made, help, hors, anytim, eddi, floyd, ranch 
#> Topic 7 Top Words:
#>       Highest Prob: will, sunday, day, live, morn, school, feel 
#>       FREX: day, live, morn, feel, fun, good, put 
#>       Lift: coffe, husband, ice, ladi, morn, park, realli 
#>       Score: feel, day, will, morn, live, sunday, put
```

## Visualizing models with `LDAvis`

Make sure you have `LDAvis` and `servr` installed.

``` r
## install LDAvis package if it's not already
if (!requireNamespace("LDAvis", quietly = TRUE)) {
  install.packages("LDAvis")
}

## install servr package if it's not already
if (!requireNamespace("servr", quietly = TRUE)) {
  install.packages("servr")
}
```

Export fitted models into interactive `LDAvis` visualizations with one
line of code

``` r
to_ldavis(model, pool.corpus, pool.dfm)
## for STM use (included in the stm package)
stm::toLDAvis(stm_model, stm_model$prep$documents)
```

![](man/figures/to_ldavis.png)

## Plotting geo-tagged tweets

Plot your tweets onto a static map

``` r
plot_tweets(mytweets, region = "USA(?!:Alaska|:Hawaii)", alpha=0.1)
```

![](man/figures/README-unnamed-chunk-14-1.png)<!-- -->

or plot the distribution of a certain hashtag onto a static map (UK data
not included)

``` r
plot_hashtag(uk_tweets, region = "UK", hashtag = "foodwaste", ignore_case=TRUE, alpha=0.2)
```

<img src="man/figures/ht_map_uk.png" width="247"/>

## Interactive maps with `leaflet`

Use scroll wheel to zoom into and out of the map. Click markets to see
tweets. Make sure you have the `leaflet` package installed.

``` r
## install leaflet package if it's not already
if (!requireNamespace("leaflet", quietly = TRUE)) {
  install.packages("leaflet")
}

cluster_tweets(mytweets)
```

![](man/figures/leaflet_us.png)


NEWS.md
# Version 0.1.5

- added `Remotes` to description to ensure `rtweet` v.0.7.0 is installed with `Twitmo`

# Version 0.1.4

-   `Twitmo` 0.1. depends on `rtweet` \< 1.0.0 now (breaking changes)
-   added linting via `lintr`
-   added `renv`
-   `pool_tweets()` now only requires a data frame with a "hashtag" and "text" column as input

# Version 0.1.3

-   Fixed an issue with `fit_stm()` where n_topics hard-coded to 7
-   Fixed an issue in `pool_tweets()` where it would expect a character vector instead of a logical when checking for duplicate tweets / retweets / quotes in the data

# Version 0.1.2

-   Fixed an issue, where `fit_stm()` default stopwords argument would lead to an error (#26)
-   Credited co-authors and advisers to the package


R/pool_tweets.R
#' Prepare Tweets for topic modeling by pooling
#' @importFrom rlang .data
#' @importFrom stats na.omit
#' @importFrom rtweet lat_lng
#' @details Pools tweets by hashtags using cosine similarity to create
#' longer pseudo-documents for better LDA estimation and creates n-gram tokens.
#' The method applies an implementation of the pooling algorithm from Mehrotra et al. 2013.
#' @description This function pools a data frame of parsed tweets into document pools.
#' @param data Data frame containing tweets and hashtags. Works with any data frame, as long as there
#' is a "text" column of type character string and a "hashtags" column with comma separated character vectors.
#' Can be obtained either by using \code{\link{load_tweets}} on a json object returned by Twitter's API v1.1 or
#' by using \code{\link[jsonlite]{stream_in}} on any json file, as long as it has a "text" and "hashtags" field.
#' If you are unsure about the requirements you may load the sample piece of data contained in the package by
#' following the example in the the example section of this help page.
#' @param remove_numbers Logical. If \code{TRUE} remove tokens that consist only of numbers,
#' but not words that start with digits, e.g. 2day. See \link[quanteda]{tokens}.
#' @param remove_punct Logical. If \code{TRUE} remove all characters in the Unicode
#' "Punctuation" [P] class, with exceptions for those used as prefixes for valid social media tags if
#' \code{preserve_tags = TRUE}. See \link[quanteda]{tokens}
#' @param remove_symbols Logical. If \code{TRUE} remove all characters in the Unicode "Symbol" [S] class.
#' @param remove_url Logical. If \code{TRUE} find and eliminate URLs beginning with http(s).
#' @param stopwords a character vector, list of character vectors, \link[quanteda]{dictionary}
#' or collocations object. See \link[quanteda]{pattern} for details.
#' Defaults to \link[stopwords:stopwords]{stopwords("english")}.
#' @param n_grams Integer vector specifying the number of elements to be concatenated in each n-gram.
#' Each element of this vector will define a n in the n-gram(s) that are produced. See \link[quanteda]{tokens_ngrams}
#' @param remove_emojis Logical. If \code{TRUE} all emojis will be removed from tweets.
#' @param remove_users Logical. If \code{TRUE} will remove all mentions of user names from documents.
#' @param remove_hashtags Logical. If \code{TRUE} will remove hashtags (not only the symbol but the hashtagged word itself) from documents.
#' @param cosine_threshold Double. Value between 0 and 1 specifying the cosine similarity threshold to be used
#' for document pooling. Tweets without a hashtag will be assigned to document (hashtag) pools
#' based upon this metric. Low thresholds will reduce topic coherence by including
#' a large number of tweets without a hashtag into the document pools. Higher thresholds will lead
#' to more coherent topics but will reduce document sizes.
#'
#' @return List with \link[quanteda]{corpus} object and \link[quanteda]{dfm} object of pooled tweets.
#' @references Mehrotra, Rishabh & Sanner, Scott & Buntine, Wray & Xie, Lexing. (2013).
#' Improving LDA Topic Models for Microblogs via Tweet Pooling and Automatic Labeling.
#' 889-892. 10.1145/2484028.2484166.
#' @seealso \link[quanteda]{tokens}, \link[quanteda]{dfm}
#'
#' @export
#' @examples
#' \dontrun{
#'
#' library(Twitmo)
#'
#' # load tweets (included in package)
#' mytweets <- load_tweets(system.file("extdata", "tweets_20191027-141233.json", package = "Twitmo"))
#'
#' pool <- pool_tweets(
#'   data = mytweets,
#'   remove_numbers = TRUE,
#'   remove_punct = TRUE,
#'   remove_symbols = TRUE,
#'   remove_url = TRUE,
#'   remove_users = TRUE,
#'   remove_hashtags = TRUE,
#'   remove_emojis = TRUE,
#'   cosine_threshold = 0.9,
#'   stopwords = "en",
#'   n_grams = 1
#' )
#' }
#'
# TODO: add meta data to dfm objects
#'
pool_tweets <- function(data,
                        remove_numbers = TRUE,
                        remove_punct = TRUE,
                        remove_symbols = TRUE,
                        remove_url = TRUE,
                        remove_emojis = TRUE,
                        remove_users = TRUE,
                        remove_hashtags = TRUE,
                        cosine_threshold = 0.9,
                        stopwords = "en",
                        n_grams = 1L) {
  quanteda::quanteda_options(pattern_hashtag = NULL, pattern_username = NULL)

  if (missing(data)) stop("Missing input data frame with no default.")

  n_grams <- as.integer(n_grams)

  stopifnot(
    "text" %in% colnames(data),
    "hashtags" %in% colnames(data),
    is.logical(remove_numbers),
    is.logical(remove_punct),
    is.logical(remove_symbols),
    is.logical(remove_url),
    is.logical(remove_emojis),
    is.logical(remove_users),
    is.logical(remove_hashtags),
    is.double(cosine_threshold),
    is.integer(n_grams),
    cosine_threshold >= 0.01 && cosine_threshold <= 1
  )

  if (cosine_threshold <= 0.2) {
    invisible(readline(prompt = "Low cosine thresholds can increase calculation time and memory usage significantly and even lead to crashes.
Press [enter] to continue or [control+c] to abort"))

    warning("Please be aware that your cosine threshold is low.
  Low cosine thresholds lead to incoherent topics due to
  document pool dillution by too tweets without hashtags.
  For coherent topics we recommend thresholds of 0.8 or higher.")
  }

  # Local stopwords + additional Twitter stopwords
  if (is.character(stopwords)) stopwords <- c(stopwords::stopwords(stopwords), "amp", "na", "rt", "via")
  if (missing(stopwords)) stopwords <- c(stopwords::stopwords("en"), "amp", "na", "rt", "via")

  cat("\n")
  cat(nrow(data), "Tweets total", sep = " ")

  # all tweets with hashtags
  tweets_w_hashtags <- data[which(nchar(data$hashtags) >= 0), ]

  # all tweets without hashtags
  tweets_no_hashtags <- data[which(is.na(data$hashtags)), ]

  cat("\n")
  cat(nrow(tweets_no_hashtags), "Tweets without hashtag", sep = " ")

  cat("\n")
  cat("Pooling", nrow(tweets_w_hashtags), "Tweets with hashtags #", sep = " ")

  hashtags.unique <- unlist(data$hashtags)
  hashtags.unique <- unique(hashtags.unique)
  hashtags.unique <- hashtags.unique[!is.na(hashtags.unique)]
  cat("\n")
  cat(length(hashtags.unique), "Unique hashtags total", sep = " ")

  cat("\n")
  cat("Begin pooling ...")

  # add single-point latitude and longitude variables to tweets data
  a <- rtweet::lat_lng(data)

  # extract emoji vector
  a$emojis <- stringr::str_extract_all(a$text, emojis_regex, simplify = FALSE)

  # remove emojis from tweets
  if (remove_emojis) {
    a$text <- stringr::str_remove_all(a$text, emojis_regex)
  }

  # remove URLs
  if (remove_url) {
    a$text <- stringr::str_replace_all(a$text, "https://t.co/[a-z,A-Z,0-9]*", "") %>%
      # remove whitespaces if tweet only consists of url
      stringr::str_trim()
  }

  # remove hashtags
  if (remove_hashtags) {
    a$text <- stringr::str_replace_all(a$text, "#[a-z,A-Z,_]*", "") %>%
      # remove whitespaces if tweet only consists of hashtags
      stringr::str_trim()
  }

  if (remove_users) {
    # remove references to usernames
    a$text <- stringr::str_replace_all(a$text, "@[a-z,A-Z,_]*", "") %>%
      # remove whitespaces if tweet only consists of usernames
      stringr::str_trim()
  }

  if (remove_numbers) {
    # remove numbers in tweets
    a$text <- stringr::str_replace_all(a$text, "[[:digit:]]", "") %>%
      # remove whitespaces if tweet only consists of digits
      stringr::str_trim()
  }

  if (remove_punct) {
    # remove all punctuation
    a$text <- stringr::str_replace_all(a$text, "[[:punct:]]", "") %>%
      # remove whitespaces if tweet only consists of punctuation
      stringr::str_trim()
  }

  # remove duplicates quoted tweets and retweets
  a <- a[a$is_quote == FALSE & a$is_retweet == FALSE, ]

  # unfold/explode twitter data by hashtags
  b <- a %>%
    tidyr::unnest(.data$hashtags)

  ## group tweets by hashtags
  c <- b %>%
    dplyr::group_by(.data$hashtags) %>%
    dplyr::summarise(text = paste0(.data$text, collapse = " "))

  # collect unique hashtags
  df <- data.frame(
    doc_id = unique(tolower(na.omit(unlist(a$hashtags)))) %>%
      seq_along() %>%
      paste0("doc", .data),
    hashtags = unique(tolower(na.omit(unlist(a$hashtags))))
  )

  # combine into docs (by hashtag)
  c$hashtags <- tolower(c$hashtags)

  # create hashtag pools by joining tweets with same hashtags
  # each tweet can be assigned to multiple pools
  # this is explicitly allowed

  d <- c %>%
    na.omit() %>%
    dplyr::group_by(.data$hashtags) %>%
    dplyr::mutate(tweets_pooled = paste0(.data$text, collapse = " ")) %>%
    dplyr::distinct(.data$hashtags, .data$tweets_pooled)

  # join document hashtag dataframe with pooled tweets dataframe
  document_hashtag_pools <- dplyr::inner_join(df, d, by = "hashtags")

  ## TF-IDF Vectorization ----
  # Calculate TF-IDF matrices for document pools ----

  # using quanteda
  doc.corpus <- quanteda::corpus(document_hashtag_pools,
    meta = document_hashtag_pools$hashtags,
    text_field = "tweets_pooled"
  )

  quanteda::docnames(doc.corpus) <- document_hashtag_pools$hashtags

  tokens.pooled <- quanteda::tokens(doc.corpus,
    what = "word",
    remove_url = remove_url,
    remove_punct = remove_punct,
    remove_symbols = remove_symbols,
    remove_numbers = remove_numbers,
    remove_separators = TRUE,
    split_hyphens = FALSE,
    include_docvars = TRUE,
    padding = FALSE
  ) %>%
    quanteda::tokens_remove(stopwords) %>%
    quanteda::tokens_ngrams(n = n_grams)


  pooled.dfm <-
    quanteda::dfm(tokens.pooled, tolower = TRUE) %>%
    quanteda::dfm_trim(.) %>%
    quanteda::dfm_tfidf(.)

  # Calculate TF-IDF matrices for unpooled tweets ----

  # tweets without hashtags
  b.nohashtag <- b[which(is.na(b$hashtags)), ]

  unpooled.corpus <- quanteda::corpus(b.nohashtag, text_field = "text")

  tokens.unpooled <- quanteda::tokens(unpooled.corpus,
    what = "word",
    remove_url = remove_url,
    remove_punct = remove_punct,
    remove_symbols = remove_symbols,
    remove_numbers = remove_numbers,
    remove_separators = TRUE,
    split_hyphens = FALSE,
    include_docvars = TRUE,
    padding = FALSE
  ) %>%
    quanteda::tokens_remove(stopwords) %>%
    quanteda::tokens_ngrams(n = n_grams)

  unpooled.dfm <-
    quanteda::dfm(tokens.unpooled, tolower = TRUE) %>%
    quanteda::dfm_trim(.) %>%
    quanteda::dfm_tfidf(.)

  # calculate cosine similarities between pooled tweets and tweets without hashtags
  h <- suppressWarnings(quanteda.textstats::textstat_simil(pooled.dfm,
    unpooled.dfm,
    margin = "documents",
    method = "cosine"
  ))

  # sample tweets using cosine threshold
  O <- as.data.frame(h)
  tt <- O[O$cosine >= cosine_threshold, ]

  # append tweets passing threshold to corresponding hashtag pools
  # skip if no tweets pass the similarity threshold
  if (!is.na(tt[1, 1])) {
    for (i in tt) {
      for (j in tt) {
        document_hashtag_pools[document_hashtag_pools["hashtags"] == as.character(i), "tweets_pooled"] <-
          paste(document_hashtag_pools[document_hashtag_pools["hashtags"] == as.character(i), "tweets_pooled"], b.nohashtag[as.character(j), "text"])
      }
    }
  }


  ## Recalculate document frequency matrices after pooling
  doc.corpus <- quanteda::corpus(document_hashtag_pools,
    meta = document_hashtag_pools$hashtags,
    text_field = "tweets_pooled"
  )

  quanteda::docnames(doc.corpus) <- document_hashtag_pools$hashtags

  # remove empty documents from corpus
  doc.corpus <- quanteda::corpus_trim(doc.corpus, what = "documents", min_ntoken = 3)

  tokens.final <- quanteda::tokens(doc.corpus,
    what = "word",
    remove_url = remove_url,
    remove_punct = remove_punct,
    remove_symbols = remove_symbols,
    remove_numbers = remove_numbers,
    remove_separators = TRUE,
    split_hyphens = FALSE,
    include_docvars = TRUE,
    padding = FALSE
  ) %>%
    quanteda::tokens_remove(stopwords) %>%
    quanteda::tokens_ngrams(n = n_grams)

  # Final pooled document frequency matrix
  pooled.dfm <- quanteda::dfm(tokens.final, tolower = TRUE)

  hashtag.freq <- a$hashtags[!is.na(a$hashtags)] %>%
    unlist() %>%
    tolower() %>%
    plyr::count() %>%
    dplyr::arrange(-.data$freq) %>%
    dplyr::as_tibble() %>%
    dplyr::rename(hashtag = .data$x, count = .data$freq)

  ret_list <- list(
    "meta" = a,
    "hashtags" = hashtag.freq,
    "tokens" = tokens.final,
    "corpus" = doc.corpus,
    "document_term_matrix" = pooled.dfm
  )
  cat("Done\n")

  return(ret_list)
}

## quiets concerns of R CMD check re: the .'s that appear in pipelines
if (getRversion() >= "2.15.1") utils::globalVariables(c("."))
